include/math/SquareRankFourTensor.hpp

```diff
@@ -209,9 +209,104 @@ class SquareRankFourTensor: public Tensor<_Scalar, 4> {
+        // For the const argument, we need the const in the template
+        //      For more info, see: https://stackoverflow.com/questions/45283468/eigen-const-tensormap
+        Eigen::TensorMap<Eigen::Tensor<const TransformationScalar, 2>> T_tensor (T.data(), T.rows(), T.cols());
+
+
+        // We will have to do four single contractions, so we specify the contraction indices
+        // Eigen3 does not document its tensor contraction clearly, so see the accepted answer on stackoverflow (https://stackoverflow.com/a/47558349/7930415):
+        //      Eigen3 does not accept a way to specify the output axes: instead, it retains the order from left to right of the axes that survive the contraction.
+        //      This means that, in order to get the right ordering of the axes, we will have to swap axes
+
+        // g(T U V W)  T^*(V R) -> a(T U R W) but we get a(T U W R)
+        Eigen::array<Eigen::IndexPair<int>, 1> contraction_pair1 = {Eigen::IndexPair<int>(2, 0)};
+        Eigen::array<int, 4> shuffle_1 {0, 1, 3, 2};
+
+        // a(T U R W)  T(W S) -> b(T U R S) and we get b(T U R S), so no shuffle is needed
+        Eigen::array<Eigen::IndexPair<int>, 1> contraction_pair2 = {Eigen::IndexPair<int>(3, 0)};
+
+        // T(U Q)  b(T U R S) -> c(T Q R S) but we get c(Q T R S)
+        Eigen::array<Eigen::IndexPair<int>, 1> contraction_pair3 = {Eigen::IndexPair<int>(0, 1)};
+        Eigen::array<int, 4> shuffle_3 {1, 0, 2, 3};
+
+        // T^*(T P)  c(T Q R S) -> g'(P Q R S) and we get g_SO(P Q R S), so no shuffle is needed
+        Eigen::array<Eigen::IndexPair<int>, 1> contraction_pair4 = {Eigen::IndexPair<int>(0, 0)};
+
+
+        // Calculate the contractions. We write this as one large contraction to
+        //  1) avoid storing intermediate contractions
+        //  2) let Eigen figure out some optimizations
+        Self g_transformed = T_tensor.conjugate().contract(T_tensor.contract(this->contract(T_tensor.conjugate(), contraction_pair1).shuffle(shuffle_1).contract(T_tensor, contraction_pair2), contraction_pair3).shuffle(shuffle_3), contraction_pair4);
+
+        *this = g_transformed;
+    }
+
+
+    /**
+     *  In-place contraction of a given matrix with this square rank-4 tensor
+     *
+     *  @tparam TransformationScalar        the type of scalar used for the transformation matrix
+     *
+     *  @param M                            the contraction matrix: contraction performed with the first index axis
+     *  @param tensor_index                 the index axis which is contracted
+     *
+     *    e.g.
+     *     if index 0 (1-st) is chosen:
+     *      T_transformed (i, j , k, l) = Sum(q) T (q j k l) * A (q, i)
+     *     if index 2 (3-rd) is chosen:
+     *      T_transformed (i, j , k, l) = Sum(q) T (i j q l) * A (q, k)
+     *
+     */
+    template <typename MultiplicationScalar = Scalar>
+    void matrixContraction(const SquareMatrix<MultiplicationScalar>& M, size_t index) {
```

Comment:
Maybe rename to `contract`? I don't think you specifically have to include `matrix` in the name.

Reply:
contract is ambiguous does not compile